make NOTE diagnostic printing in APT optional

reegnz opened the pull request:
Since javac doesn't support hiding only NOTE diagnostic messages,
you have to hide them together with WARN. That's usually not what
you want however, since compiler WARN messages might point out
bugs in your code, so you would want to leave them on.

Solution is to not print NOTE by default and enable it with an
additional compiler arg with -A:
http://docs.oracle.com/javase/7/docs/technotes/tools/windows/javac.html#options

With this change that means you can pass -Aquerydsl.logInfo=true
will
enable NOTE messages, but by default they are not logged.

resolves #2032

timowest commented on querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java:
```diff
@@ -539,6 +540,24 @@ public SourceVersion getSupportedSourceVersion() {
+        return optionKeys;
+    }
+
+    private boolean shouldLogInfo() {
```
put this into a field instead of resolving for each call

reegnz commented:
Hi, any updates on this?

Shredder121 commented on querydsl-apt/src/main/java/com/querydsl/apt/AbstractQuerydslProcessor.java:
```diff
@@ -51,6 +50,8 @@
 
     public static final Boolean ALLOW_OTHER_PROCESSORS_TO_CLAIM_ANNOTATIONS = Boolean.FALSE;
 
+    private static final String LOG_PROPERTY_KEY = "querydsl.apt.log.info";
```
Could you move this to `APTOptions`.

reegnz replied:
Sure, will do.

Shredder121 commented:
Looks fine to me, just one comment.

reegnz commented:
Moved property to APTOptions, squashed everything into a single commit.
What about tests? Should I have any? (To be honest, I am not too familiar with testing annotation processors...)

Shredder121 commented:
>What about tests? Should I have any?

Yes, that's an annoying issue.
You could add a new subclass of `AbstractProcessorTest` that replaces the `System.setOut` to a `ByteArrayOutputStream`.
Make sure to override `getAPTOptions()` with the new property you want to test.

Then after `process()` is called, verify the stream is empty.

reegnz commented:
Added some sanity tests:
without args note logging is disabled
with -Aquerydsl.logInfo=false note logging is disabled
with -Aquerydsl.logInfo=true note logging is enabled

Shredder121 commented:
Looks good to me!
@querydsl/developers What do you all think?

reegnz commented:
Any update @querydsl/developers?

timowest commented on querydsl-apt/src/test/java/com/querydsl/apt/NoteTest.java:
```diff
@@ -0,0 +1,63 @@
+
+public class NoteTest extends AbstractProcessorTest {
+
+        private Collection<String> aptOptions;
```
indentation

timowest commented on querydsl-apt/src/test/java/com/querydsl/apt/AbstractProcessorTest.java:
```diff
@@ -77,6 +77,14 @@ protected void compile(Class<? extends AbstractProcessor> processorClass, List<S
     }
 
+    protected ByteArrayOutputStream getStdOut() {
+            return new ByteArrayOutputStream();
```
indentation

reegnz commented:
Fixed indentation, rebased on latest HEAD of master branch.

reegnz commented:
I would also look into the checkstyle rule on why it allowed that 8 spaces instead of 4 for indentation on the tests.

Shredder121 commented:
That's.. at least interesting, I know we don't have the same checks for main and test source, but code style should probably be in both. (As far as code generation cooperates with those rules)
